File: external/google-tv-pairing-protocol/java/src/com/google/polo/pairing/PoloUtil.java
/*
 * Copyright (C) 2009 Google Inc.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.polo.pairing;

import com.google.polo.exception.PoloException;

import java.math.BigInteger;
import java.security.cert.Certificate;

import javax.net.ssl.SSLPeerUnverifiedException;
import javax.net.ssl.SSLSession;

/**
 * Utility methods of general usefulness to the Polo library.
 */
public class PoloUtil {

  /**
   * Returns the peer {@link Certificate} for an {@link SSLSession}.
   * 
   * @throws PoloException  if the peer certificate could not be obtained
   *                        from the {@link SSLSession}.
   * @return                the {@link Certificate} of the peer
   */
  public static Certificate getPeerCert(SSLSession session)
      throws PoloException {
    try {      
      // Peer certificate
      Certificate[] certs = session.getPeerCertificates();
      if (certs == null || certs.length < 1) {
        throw new PoloException("No peer certificate.");
      }
      return certs[0];
    } catch (SSLPeerUnverifiedException e) {
      throw new PoloException(e);
    }
  }
  
  /**
   * Return the local {@link Certificate} for an {@link SSLSession}.
   * 
   * @throws PoloException  if the local certificate could not be obtained
   *                        from the {@link SSLSession}
   * @return                the {@link Certificate} of the peer
   */
  public static Certificate getLocalCert(SSLSession session)
      throws PoloException {
    Certificate[] certs = session.getLocalCertificates();
    if (certs == null || certs.length < 1) {
      throw new PoloException("No local certificate.");
    }
    return certs[0];
  }
  
  /**
   * Converts an array of bytes to a string of hexadecimal characters.
   * Leading null bytes are preserved in the output.
   * <p>
   * The input byte stream is assumed to be a positive, two's complement
   * representation of an integer.  The return value is the hexadecimal string
   * representation of this value.
   * 
   * @param bytes  the bytes to convert
   * @return       the string representation
   */
  public static String bytesToHexString(byte[] bytes) {
    if (bytes == null || bytes.length == 0) {
      return "";
    }
    BigInteger bigint = new BigInteger(1, bytes);
    int formatLen = bytes.length * 2;
    return String.format("%0" + formatLen + "x", bigint);
  }
  
  /**
   * Converts a string of hex characters to a byte array.
   * 
   * @param hexstr  the string of hex characters
   * @return        a byte array representation
   */
  public static byte[] hexStringToBytes(String hexstr) {
    if (hexstr == null || hexstr.length() == 0 || (hexstr.length() % 2) != 0) {
      throw new IllegalArgumentException("Bad input string.");
    }
    
    byte[] result = new byte[hexstr.length() / 2];
    for (int i=0; i < result.length; i++) {
      result[i] = (byte) Integer.parseInt(hexstr.substring(2 * i, 2 * (i + 1)),
          16);
    }
    return result;
  }
  
  /**
   * Converts an integer value to the big endian 4-byte representation.
   */
  public static final byte[] intToBigEndianIntBytes(int intVal) {
    byte[] outBuf = new byte[4];
    outBuf[0] = (byte)((intVal >> 24) & 0xff);
    outBuf[1] = (byte)((intVal >> 16) & 0xff);
    outBuf[2] = (byte)((intVal >> 8) & 0xff);
    outBuf[3] = (byte)(intVal & 0xff);
    return outBuf;
  }

  /**
   * Converts a 4-byte array of bytes to an unsigned long value.
   */
  public static final long intBigEndianBytesToLong(byte[] input) {
    assert (input.length == 4);
    long ret = (long)(input[0]) & 0xff;
    ret <<= 8;
    ret |= (long)(input[1]) & 0xff;
    ret <<= 8;
    ret |= (long)(input[2]) & 0xff;
    ret <<= 8;
    ret |= (long)(input[3]) & 0xff;
    return ret;
  }

}

File: external/google-tv-pairing-protocol/cpp/src/polo/util/poloutil.cc
// Copyright 2012 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "polo/util/poloutil.h"

#include <openssl/rand.h>

namespace polo {
namespace util {

const std::string PoloUtil::BytesToHexString(const uint8_t* bytes,
                                             size_t length) {
  // Use OpenSSL BigNum functions to perform the conversion.
  BIGNUM* bn = BN_bin2bn(bytes, length, NULL);
  char* hex = BN_bn2hex(bn);
  std::string hex_string(hex);
  BN_free(bn);
  OPENSSL_free(hex);
  return hex_string;
}

const size_t PoloUtil::HexStringToBytes(const std::string hex_string,
                                        uint8_t*& bytes) {
  // Use OpenSSL BigNum functions to perform the conversion.
  BIGNUM* bn = NULL;
  BN_hex2bn(&bn, hex_string.c_str());
  int length = BN_num_bytes(bn);
  bytes = new uint8_t[length];
  BN_bn2bin(bn, &bytes[0]);
  BN_free(bn);
  return length;
}

const void PoloUtil::IntToBigEndianBytes(uint32_t value,
                                         uint8_t*& bytes) {
  // Use OpenSSL BigNum functions to perform the conversion.
  BIGNUM* bn = BN_new();
  BN_add_word(bn, value);

  // Initialize the array to 0 so there will be leading null bytes if the
  // number is less than 4 bytes long.
  bytes = new uint8_t[4];
  for (int i = 0; i < 4; i++) {
    bytes[i] = 0;
  }

  int length = BN_num_bytes(bn);
  BN_bn2bin(bn, &bytes[4 - length]);
  BN_free(bn);
}

const uint32_t PoloUtil::BigEndianBytesToInt(
    const uint8_t* bytes) {
  // Use OpenSSL BigNum functions to perform the conversion.
  BIGNUM* bn = BN_bin2bn(bytes, 4, NULL);
  BN_ULONG value = bn->d[0];
  BN_free(bn);
  return value;
}

uint8_t* PoloUtil::GenerateRandomBytes(size_t length) {
  // Use the OpenSSL library to generate the random byte array. The RAND_bytes
  // function is guaranteed to provide secure random bytes.
  uint8_t* buffer = new uint8_t[length];
  if (RAND_bytes(buffer, length)) {
    return buffer;
  } else {
    delete buffer;
    return NULL;
  }
}

}  // namespace util
}  // namespace polo
